lib/journaled/connection.rb

```diff
@@ -0,0 +1,48 @@
+module Journaled
+  module Connection
+    class << self
+      def available?
+        Journaled.transactional_batching_enabled && transaction_joinable?
+      end
+
+      def stage!(event)
+        raise TransactionSafetyError, <<~MSG unless transaction_joinable?
+          Transaction not available! By default, journaled event batching requires an open database transaction.
+        MSG
+
+        connection._journaled_transaction_handler.staged_events << event
+      end
+
+      private
+
+      def transaction_joinable?
+        connection.transaction_open? && connection._journaled_transaction_handler.joinable?
+      end
+
+      def connection
+        if Journaled.queue_adapter.in? %w(delayed delayed_job)
+          Delayed::Job.connection
+        elsif Journaled.queue_adapter == 'good_job'
+          GoodJob::BaseRecord.connection
+        elsif Journaled.queue_adapter == 'que'
+          Que::ActiveRecord::Model.connection
+        elsif Journaled.queue_adapter == 'test' && Rails.env.test?
+          ActiveRecord::Base.connection
+        else
+          raise "Unsupported adapter: #{Journaled.queue_adapter}"
+        end
+      end
+    end
+
+    module TestOnlyBehaviors
+      def transaction_joinable?
+        # Transactional fixtures wrap all tests in an outer, non-joinable transaction:
+        super && (connection.open_transactions > 1 || connection.current_transaction.joinable?)
```

Comment:
This ensures that when `config.use_transactional_tests = true` on any tests, we don't stage events onto that outermost test transaction. Rails marks that particular transaction as non-joinable (which means that any nested transactions will behave as if they are the outer-most transactions, and will run their own callbacks at the right time), but if open_transactions > 1 we'll shortcut that check and return true, because we want to remain as close to production behavior as possible.